lib/tasks/after_tasks.rake

```diff
@@ -740,6 +740,39 @@ namespace :After do
     WorkIndexer.create_mapping
   end
 
+  desc "Fix tags with leading or trailing spaces"
+  task(fix_tags_with_leading_or_trailing_spaces: :environment) do
+    total_tags = Tag.count
+    total_batches = (total_tags + 999) / 1000
+    puts "Inspecting #{total_tags} tags in #{total_batches} batches"
+
+    report_string = "Tag ID,Old tag name,New tag name\n"
+    Tag.find_in_batches.with_index do |batch, index|
+      batch_number = index + 1
+      progress_msg = "Batch #{batch_number} of #{total_batches} complete"
+
+      batch.each do |tag|
+        if tag.name != tag.name.squish
+          old_tag_name = tag.name
+
+          leading_spaces = old_tag_name.match(/^(\s+)/)
```

Comment:
Since the issue mentioned looking at how [`squish`](https://apidock.com/rails/v5.2.3/String/squish%21) detects spaces... is `\s` exactly the same as `[[:space:]]`?

I honestly do not know the answer and I found the use of "etc" in the [Regexp class documentation](https://ruby-doc.org/core-3.0.1/Regexp.html#class-Regexp-label-Character+Classes) entirely unhelpful. I'm guessing yes based on [this page on POSIX bracket expressions](https://www.regular-expressions.info/posixbrackets.html), but I'd rather ask a silly question that assume wrong.

Reply:
I was wondering *why* the issue mentioned that. (I still don't know.)

The issue also only mentions "leading and trailing" spaces, but not any accidentally duplicated spaces in the middle of the tag, which `squish` also compresses.

Do we in fact want to squish the middle spaces as well? Or is using squish just a heuristic to narrow down which tags might have blanks in the front and in the back, even if it might catch tags that only have extra spaces in the middle, but that we don't want to mess with?

Reply:
They are not the same:
```
2.6.5 :038 > "\u00A0\u2002\u2003\u202F\u205F".gsub(/\s/, "_")
 => "     " 
2.6.5 :039 > "\u00A0\u2002\u2003\u202F\u205F".gsub(/[[:space:]]/, "_")
 => "_____" 
```
The difference in their behavior for various unicode space characters is what caused the issue in the first place, I believe. The tag code has been [squishing tag names before validation since 2008](https://github.com/otwcode/otwarchive/blob/b63fce67a3975ec0614bb535c75d76c8dafbda71/app/models/tag.rb#L210-L213), but with the update to Rails 4, the behavior of `squish` changed (because it switched from `\s` to `[[:space:]]`). That means that there may be some legacy tags from before the Rails upgrade that have `tag.name != tag.name.squish`, since various space characters (like the non-breaking space) didn't used to be treated as space characters when calling `squish`.

And yeah, it should also squish sequences of `[[:space:]]` characters in the middle of the name (or replace them with underscores as needed). All tags need to have `tag.name == tag.name.squish` to prevent issues in the future.